// angular-cesium/esm2015/lib/angular-cesium/components/ac-array-desc/ac-array-desc.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChildren, Input, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
import * as _get from 'lodash.get';
import { AcLayerComponent } from '../ac-layer/ac-layer.component';
import { LayerService } from '../../services/layer-service/layer-service.service';
import { BasicDesc } from '../../services/basic-desc/basic-desc.service';
/**
 *  This is component represents an array under `ac-layer`.
 *  The element must be a child of ac-layer element.
 *  + acFor `{string}` - get the tracked array and entityName (see the example).
 *  + idGetter `{Function}` - a function that gets the id for a given element in the array -should be defined for maximum performance.
 *  + show `{boolean}` - show/hide array's entities.
 *
 *  __Usage :__
 *  ```
 * <ac-layer acFor="let track of tracks$" [show]="show" [context]="this" [store]="true">
 *  <ac-array-desc acFor="let arrayItem of track.array" [idGetter]="trackArrayIdGetter">
 *    <ac-array-desc acFor="let innerArrayItem of arrayItem.innerArray" [idGetter]="trackArrayIdGetter">
 *      <ac-point-desc props="{
 *        position: innerArrayItem.pos,
 *        pixelSize: 10,
 *        color: getTrackColor(track),
 *        outlineColor: Cesium.Color.BLUE,
 *        outlineWidth: 1
 *      }">
 *      </ac-point-desc>
 *    </ac-array-desc>
 *  </ac-array-desc>
 * </ac-layer>
 *  ```
 */
export class AcArrayDescComponent {
    /**
     * @param {?} layerService
     * @param {?} cd
     */
    constructor(layerService, cd) {
        this.layerService = layerService;
        this.cd = cd;
        this.show = true;
        this.entitiesMap = new Map();
        this.id = 0;
        this.acForRgx = /^let\s+.+\s+of\s+.+$/;
        this.arrayObservable$ = new Subject();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes['acFor'].firstChange) {
            /** @type {?} */
            const acForString = changes['acFor'].currentValue;
            if (!this.acForRgx.test(acForString)) {
                throw new Error(`ac-layer: Invalid [acFor] syntax. Expected: [acFor]="let item of observable" .Instead received: ${acForString}`);
            }
            /** @type {?} */
            const acForArr = changes['acFor'].currentValue.split(' ');
            this.arrayPath = acForArr[3];
            this.entityName = acForArr[1];
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.layer) {
            this.layer.getLayerService().cache = false;
        }
        this.layerServiceSubscription = this.layerService.layerUpdates().subscribe(() => {
            this.cd.detectChanges();
        });
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.layerService.context['arrayObservable$'] = this.arrayObservable$;
        this.layerService.registerDescription(this);
        this.basicDescs._results.forEach((component) => {
            component.setLayerService(this.layer.getLayerService());
        });
        this.arrayDescs._results.splice(0, 1);
        this.arrayDescs._results.forEach((component) => {
            this.layerService.unregisterDescription(component);
            this.layer.getLayerService().registerDescription(component);
            component.layerService = this.layer.getLayerService();
            component.setLayerService(this.layer.getLayerService());
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.layerServiceSubscription) {
            this.layerServiceSubscription.unsubscribe();
        }
    }
    /**
     * @param {?} layerService
     * @return {?}
     */
    setLayerService(layerService) {
        this.layerService = layerService;
    }
    /**
     * @param {?} context
     * @param {?} id
     * @param {?} contextEntity
     * @return {?}
     */
    draw(context, id, contextEntity) {
        /** @type {?} */
        const get = _get;
        /** @type {?} */
        const entitiesArray = get(context, this.arrayPath);
        if (!entitiesArray) {
            return;
        }
        /** @type {?} */
        const previousEntitiesIdArray = this.entitiesMap.get(id);
        /** @type {?} */
        const entitiesIdArray = [];
        this.entitiesMap.set(id, entitiesIdArray);
        entitiesArray.forEach((item, index) => {
            this.layerService.context[this.entityName] = item;
            /** @type {?} */
            const arrayItemId = this.generateCombinedId(id, item, index);
            entitiesIdArray.push(arrayItemId);
            this.layer.update(contextEntity, arrayItemId);
        });
        if (previousEntitiesIdArray) {
            /** @type {?} */
            const entitiesToRemove = this.idGetter ?
                previousEntitiesIdArray.filter((entityId) => entitiesIdArray.indexOf(entityId) < 0) :
                previousEntitiesIdArray;
            if (entitiesToRemove) {
                entitiesToRemove.forEach((entityId) => this.layer.remove(entityId));
            }
        }
    }
    /**
     * @param {?} id
     * @return {?}
     */
    remove(id) {
        /** @type {?} */
        const entitiesIdArray = this.entitiesMap.get(id);
        if (entitiesIdArray) {
            entitiesIdArray.forEach((entityId) => this.layer.remove(entityId));
        }
        this.entitiesMap.delete(id);
    }
    /**
     * @return {?}
     */
    removeAll() {
        this.layer.removeAll();
        this.entitiesMap.clear();
    }
    /**
     * @return {?}
     */
    getAcForString() {
        return `let ${this.entityName + '___temp'} of arrayObservable$`;
    }
    /**
     * @private
     * @param {?} entityId
     * @param {?} arrayItem
     * @param {?} index
     * @return {?}
     */
    generateCombinedId(entityId, arrayItem, index) {
        /** @type {?} */
        let arrayItemId;
        if (this.idGetter) {
            arrayItemId = this.idGetter(arrayItem, index);
        }
        else {
            arrayItemId = (this.id++) % Number.MAX_SAFE_INTEGER;
        }
        return entityId + arrayItemId;
    }
}
AcArrayDescComponent.decorators = [
    { type: Component, args: [{
                selector: 'ac-array-desc',
                template: `
    <ac-layer #layer [acFor]="getAcForString()"
              [context]="layerService.context"
              [options]="layerService.options"
              [show]="layerService.show && show"
              [zIndex]="layerService.zIndex">
      <ng-content #content></ng-content>
    </ac-layer>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
AcArrayDescComponent.ctorParameters = () => [
    { type: LayerService },
    { type: ChangeDetectorRef }
];
AcArrayDescComponent.propDecorators = {
    acFor: [{ type: Input }],
    idGetter: [{ type: Input }],
    show: [{ type: Input }],
    layer: [{ type: ViewChild, args: ['layer', { static: true },] }],
    basicDescs: [{ type: ContentChildren, args: [BasicDesc, { descendants: false },] }],
    arrayDescs: [{ type: ContentChildren, args: [AcArrayDescComponent, { descendants: false },] }]
};
if (false) {
    /** @type {?} */
    AcArrayDescComponent.prototype.acFor;
    /** @type {?} */
    AcArrayDescComponent.prototype.idGetter;
    /** @type {?} */
    AcArrayDescComponent.prototype.show;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.layer;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.basicDescs;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.arrayDescs;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.entitiesMap;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.layerServiceSubscription;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.id;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.acForRgx;
    /** @type {?} */
    AcArrayDescComponent.prototype.entityName;
    /** @type {?} */
    AcArrayDescComponent.prototype.arrayPath;
    /** @type {?} */
    AcArrayDescComponent.prototype.arrayObservable$;
    /** @type {?} */
    AcArrayDescComponent.prototype.layerService;
    /**
     * @type {?}
     * @private
     */
    AcArrayDescComponent.prototype.cd;
}
