// angular-cesium/esm5/lib/angular-cesium-widgets/models/edit-point.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AcEntity } from '../../angular-cesium/models/ac-entity';
var EditPoint = /** @class */ (function (_super) {
    tslib_1.__extends(EditPoint, _super);
    function EditPoint(entityId, position, pointProps, virtualPoint) {
        if (virtualPoint === void 0) { virtualPoint = false; }
        var _this = _super.call(this) || this;
        _this._show = true;
        _this.editedEntityId = entityId;
        _this.position = position;
        _this.id = _this.generateId();
        _this.pointProps = tslib_1.__assign({}, pointProps);
        _this._virtualEditPoint = virtualPoint;
        return _this;
    }
    Object.defineProperty(EditPoint.prototype, "show", {
        get: /**
         * @return {?}
         */
        function () {
            return this._show;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._show = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EditPoint.prototype, "props", {
        get: /**
         * @return {?}
         */
        function () {
            return this.pointProps;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.pointProps = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    EditPoint.prototype.isVirtualEditPoint = /**
     * @return {?}
     */
    function () {
        return this._virtualEditPoint;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    EditPoint.prototype.setVirtualEditPoint = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._virtualEditPoint = value;
    };
    /**
     * @return {?}
     */
    EditPoint.prototype.getEditedEntityId = /**
     * @return {?}
     */
    function () {
        return this.editedEntityId;
    };
    /**
     * @return {?}
     */
    EditPoint.prototype.getPosition = /**
     * @return {?}
     */
    function () {
        return this.position.clone();
    };
    /**
     * @return {?}
     */
    EditPoint.prototype.getPositionCallbackProperty = /**
     * @return {?}
     */
    function () {
        return new Cesium.CallbackProperty(this.getPosition.bind(this), false);
    };
    /**
     * @param {?} position
     * @return {?}
     */
    EditPoint.prototype.setPosition = /**
     * @param {?} position
     * @return {?}
     */
    function (position) {
        this.position.x = position.x;
        this.position.y = position.y;
        this.position.z = position.z;
    };
    /**
     * @return {?}
     */
    EditPoint.prototype.getId = /**
     * @return {?}
     */
    function () {
        return this.id;
    };
    /**
     * @private
     * @return {?}
     */
    EditPoint.prototype.generateId = /**
     * @private
     * @return {?}
     */
    function () {
        return 'edit-point-' + EditPoint.counter++;
    };
    EditPoint.counter = 0;
    return EditPoint;
}(AcEntity));
export { EditPoint };
if (false) {
    /** @type {?} */
    EditPoint.counter;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.id;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.editedEntityId;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.position;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype._virtualEditPoint;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.pointProps;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype._show;
}
