// angular-cesium/esm5/lib/angular-cesium/components/ac-context-menu-wrapper/ac-context-menu-wrapper.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ComponentFactoryResolver, ViewChild, ViewContainerRef } from '@angular/core';
import { ContextMenuService } from '../../services/context-menu/context-menu.service';
/**
 * This component is used to inject the component that is passed to the ContextMenuService when opening a context menu.
 * It shouldn't be used directly.
 *
 * usage:
 * ```typescript
 * // We want to open the context menu on mouse right click.
 * // Register to mouse right click with the MapEventsManager
 * this.mapEventsManager.register({ event: CesiumEvent.RIGHT_CLICK, pick: PickOptions.NO_PICK })
 *    .subscribe(event => {
 *       const position = this.coordinateConverter.screenToCartesian3(event.movement.endPosition, true);
 *       if (!position) {
 *         return;
 *       }
 *       // Open the context menu on the position that was clicked and pass some data to ContextMenuComponent.
 *       this.contextMenuService.open(
 *         ContextMenuComponent,
 *         position,
 *         { data: { items: ['New Track', 'Change Map', 'Context Menu', 'Do Something'] } }
 *       )
 *    });
 *
 * ```
 */
var AcContextMenuWrapperComponent = /** @class */ (function () {
    function AcContextMenuWrapperComponent(contextMenuService, cd, componentFactoryResolver) {
        this.contextMenuService = contextMenuService;
        this.cd = cd;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    /**
     * @return {?}
     */
    AcContextMenuWrapperComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.contextMenuChangeSubscription =
            this.contextMenuService.contextMenuChangeNotifier.subscribe(function () { return _this.cd.detectChanges(); });
        this.contextMenuOpenSubscription =
            this.contextMenuService.onOpen.subscribe(function () {
                /** @type {?} */
                var componentFactory = _this.componentFactoryResolver.resolveComponentFactory((/** @type {?} */ (_this.contextMenuService.content)));
                _this.viewContainerRef.clear();
                /** @type {?} */
                var componentRef = _this.viewContainerRef.createComponent(componentFactory);
                ((/** @type {?} */ (componentRef.instance))).data = _this.contextMenuService.options.data;
                _this.cd.detectChanges();
            });
    };
    /**
     * @return {?}
     */
    AcContextMenuWrapperComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.contextMenuChangeSubscription) {
            this.contextMenuChangeSubscription.unsubscribe();
        }
        if (this.contextMenuOpenSubscription) {
            this.contextMenuOpenSubscription.unsubscribe();
        }
    };
    AcContextMenuWrapperComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ac-context-menu-wrapper',
                    template: "\n    <ac-html *ngIf=\"contextMenuService.showContextMenu\" [props]=\"{position: contextMenuService.position}\">\n      <div #contextMenuContainer></div>\n    </ac-html>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    AcContextMenuWrapperComponent.ctorParameters = function () { return [
        { type: ContextMenuService },
        { type: ChangeDetectorRef },
        { type: ComponentFactoryResolver }
    ]; };
    AcContextMenuWrapperComponent.propDecorators = {
        viewContainerRef: [{ type: ViewChild, args: ['contextMenuContainer', { static: false, read: ViewContainerRef },] }]
    };
    return AcContextMenuWrapperComponent;
}());
export { AcContextMenuWrapperComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AcContextMenuWrapperComponent.prototype.contextMenuChangeSubscription;
    /**
     * @type {?}
     * @private
     */
    AcContextMenuWrapperComponent.prototype.contextMenuOpenSubscription;
    /** @type {?} */
    AcContextMenuWrapperComponent.prototype.viewContainerRef;
    /** @type {?} */
    AcContextMenuWrapperComponent.prototype.contextMenuService;
    /**
     * @type {?}
     * @private
     */
    AcContextMenuWrapperComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    AcContextMenuWrapperComponent.prototype.componentFactoryResolver;
}
