// angular-cesium/esm2015/lib/angular-cesium-widgets/models/edit-polyline.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AcEntity } from '../../angular-cesium/models/ac-entity';
export class EditPolyline extends AcEntity {
    /**
     * @param {?} entityId
     * @param {?} startPosition
     * @param {?} endPosition
     * @param {?=} polylineProps
     */
    constructor(entityId, startPosition, endPosition, polylineProps) {
        super();
        this.editedEntityId = entityId;
        this.id = this.generateId();
        this.positions = [startPosition, endPosition];
        this._polylineProps = Object.assign({}, polylineProps);
    }
    /**
     * @return {?}
     */
    get props() {
        return this._polylineProps;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set props(value) {
        this._polylineProps = value;
    }
    /**
     * @return {?}
     */
    getEditedEntityId() {
        return this.editedEntityId;
    }
    /**
     * @return {?}
     */
    getPositions() {
        return this.positions.map(p => p.clone());
    }
    /**
     * @return {?}
     */
    getPositionsCallbackProperty() {
        return new Cesium.CallbackProperty(this.getPositions.bind(this), false);
    }
    /**
     * @return {?}
     */
    validatePositions() {
        return this.positions[0] !== undefined && this.positions[1] !== undefined;
    }
    /**
     * @return {?}
     */
    getStartPosition() {
        return this.positions[0];
    }
    /**
     * @return {?}
     */
    getEndPosition() {
        return this.positions[1];
    }
    /**
     * @param {?} position
     * @return {?}
     */
    setStartPosition(position) {
        this.positions[0] = position;
    }
    /**
     * @param {?} position
     * @return {?}
     */
    setEndPosition(position) {
        this.positions[1] = position;
    }
    /**
     * @return {?}
     */
    getId() {
        return this.id;
    }
    /**
     * @private
     * @return {?}
     */
    generateId() {
        return 'edit-polyline-' + EditPolyline.counter++;
    }
}
EditPolyline.counter = 0;
if (false) {
    /** @type {?} */
    EditPolyline.counter;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype.editedEntityId;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype.id;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype.positions;
    /**
     * @type {?}
     * @private
     */
    EditPolyline.prototype._polylineProps;
}
