// angular-cesium/esm2015/lib/angular-cesium-widgets/models/edit-point.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AcEntity } from '../../angular-cesium/models/ac-entity';
export class EditPoint extends AcEntity {
    /**
     * @param {?} entityId
     * @param {?} position
     * @param {?=} pointProps
     * @param {?=} virtualPoint
     */
    constructor(entityId, position, pointProps, virtualPoint = false) {
        super();
        this._show = true;
        this.editedEntityId = entityId;
        this.position = position;
        this.id = this.generateId();
        this.pointProps = Object.assign({}, pointProps);
        this._virtualEditPoint = virtualPoint;
    }
    /**
     * @return {?}
     */
    get show() {
        return this._show;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set show(value) {
        this._show = value;
    }
    /**
     * @return {?}
     */
    get props() {
        return this.pointProps;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set props(value) {
        this.pointProps = value;
    }
    /**
     * @return {?}
     */
    isVirtualEditPoint() {
        return this._virtualEditPoint;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    setVirtualEditPoint(value) {
        this._virtualEditPoint = value;
    }
    /**
     * @return {?}
     */
    getEditedEntityId() {
        return this.editedEntityId;
    }
    /**
     * @return {?}
     */
    getPosition() {
        return this.position.clone();
    }
    /**
     * @return {?}
     */
    getPositionCallbackProperty() {
        return new Cesium.CallbackProperty(this.getPosition.bind(this), false);
    }
    /**
     * @param {?} position
     * @return {?}
     */
    setPosition(position) {
        this.position.x = position.x;
        this.position.y = position.y;
        this.position.z = position.z;
    }
    /**
     * @return {?}
     */
    getId() {
        return this.id;
    }
    /**
     * @private
     * @return {?}
     */
    generateId() {
        return 'edit-point-' + EditPoint.counter++;
    }
}
EditPoint.counter = 0;
if (false) {
    /** @type {?} */
    EditPoint.counter;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.id;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.editedEntityId;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.position;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype._virtualEditPoint;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype.pointProps;
    /**
     * @type {?}
     * @private
     */
    EditPoint.prototype._show;
}
