// angular-cesium/esm2015/lib/angular-cesium-widgets/models/label-props.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function LabelStyle() { }
if (false) {
    /** @type {?|undefined} */
    LabelStyle.prototype.show;
    /** @type {?|undefined} */
    LabelStyle.prototype.font;
    /** @type {?|undefined} */
    LabelStyle.prototype.style;
    /** @type {?|undefined} */
    LabelStyle.prototype.fillColor;
    /** @type {?|undefined} */
    LabelStyle.prototype.outlineColor;
    /** @type {?|undefined} */
    LabelStyle.prototype.backgroundColor;
    /** @type {?|undefined} */
    LabelStyle.prototype.backgroundPadding;
    /** @type {?|undefined} */
    LabelStyle.prototype.showBackground;
    /** @type {?|undefined} */
    LabelStyle.prototype.scale;
    /** @type {?|undefined} */
    LabelStyle.prototype.distanceDisplayCondition;
    /** @type {?|undefined} */
    LabelStyle.prototype.heightReference;
    /** @type {?|undefined} */
    LabelStyle.prototype.horizontalOrigin;
    /** @type {?|undefined} */
    LabelStyle.prototype.eyeOffset;
    /** @type {?|undefined} */
    LabelStyle.prototype.position;
    /** @type {?|undefined} */
    LabelStyle.prototype.pixelOffset;
    /** @type {?|undefined} */
    LabelStyle.prototype.pixelOffsetScaleByDistance;
    /** @type {?|undefined} */
    LabelStyle.prototype.outlineWidth;
    /** @type {?|undefined} */
    LabelStyle.prototype.scaleByDistance;
    /** @type {?|undefined} */
    LabelStyle.prototype.translucencyByDistance;
    /** @type {?|undefined} */
    LabelStyle.prototype.verticalOrigin;
}
/**
 * @record
 */
export function LabelProps() { }
if (false) {
    /** @type {?} */
    LabelProps.prototype.text;
    /** @type {?|undefined} */
    LabelProps.prototype.show;
    /** @type {?|undefined} */
    LabelProps.prototype.font;
    /** @type {?|undefined} */
    LabelProps.prototype.style;
    /** @type {?|undefined} */
    LabelProps.prototype.fillColor;
    /** @type {?|undefined} */
    LabelProps.prototype.outlineColor;
    /** @type {?|undefined} */
    LabelProps.prototype.backgroundColor;
    /** @type {?|undefined} */
    LabelProps.prototype.backgroundPadding;
    /** @type {?|undefined} */
    LabelProps.prototype.showBackground;
    /** @type {?|undefined} */
    LabelProps.prototype.scale;
    /** @type {?|undefined} */
    LabelProps.prototype.distanceDisplayCondition;
    /** @type {?|undefined} */
    LabelProps.prototype.heightReference;
    /** @type {?|undefined} */
    LabelProps.prototype.horizontalOrigin;
    /** @type {?|undefined} */
    LabelProps.prototype.eyeOffset;
    /** @type {?|undefined} */
    LabelProps.prototype.position;
    /** @type {?|undefined} */
    LabelProps.prototype.pixelOffset;
    /** @type {?|undefined} */
    LabelProps.prototype.pixelOffsetScaleByDistance;
    /** @type {?|undefined} */
    LabelProps.prototype.outlineWidth;
    /** @type {?|undefined} */
    LabelProps.prototype.scaleByDistance;
    /** @type {?|undefined} */
    LabelProps.prototype.translucencyByDistance;
    /** @type {?|undefined} */
    LabelProps.prototype.verticalOrigin;
    /** @type {?|undefined} */
    LabelProps.prototype.disableDepthTestDistance;
}
/** @type {?} */
export const defaultLabelProps = {
    backgroundColor: new Cesium.Color(0.165, 0.165, 0.165, 0.7),
    backgroundPadding: new Cesium.Cartesian2(25, 20),
    distanceDisplayCondition: undefined,
    fillColor: Cesium.Color.WHITE,
    font: '30px sans-serif',
    heightReference: Cesium.HeightReference.NONE,
    horizontalOrigin: Cesium.HorizontalOrigin.LEFT,
    outlineColor: Cesium.Color.BLACK,
    outlineWidth: 1.0,
    pixelOffset: Cesium.Cartesian2.ZERO,
    pixelOffsetScaleByDistance: undefined,
    scale: 1.0,
    scaleByDistance: undefined,
    show: true,
    showBackground: false,
    style: Cesium.LabelStyle.FILL,
    text: '',
    translucencyByDistance: undefined,
    verticalOrigin: Cesium.VerticalOrigin.BASELINE,
    disableDepthTestDistance: Number.POSITIVE_INFINITY,
};
