// angular-cesium/esm5/lib/cesium-extender/collections/html.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HtmlPrimitive } from '../primitives';
var HtmlCollection = /** @class */ (function () {
    function HtmlCollection() {
        this._collection = [];
    }
    Object.defineProperty(HtmlCollection.prototype, "length", {
        get: /**
         * @return {?}
         */
        function () {
            return this._collection.length;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} index
     * @return {?}
     */
    HtmlCollection.prototype.get = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return this._collection[index];
    };
    /**
     * @param {?} options
     * @return {?}
     */
    HtmlCollection.prototype.add = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        /** @type {?} */
        var html = new HtmlPrimitive(options, this);
        this._collection.push(html);
        return html;
    };
    /**
     * @param {?} html
     * @return {?}
     */
    HtmlCollection.prototype.remove = /**
     * @param {?} html
     * @return {?}
     */
    function (html) {
        /** @type {?} */
        var index = this._collection.indexOf(html);
        if (index === (-1)) {
            return false;
        }
        this._collection[index].remove();
        this._collection.splice(index, 1);
        return true;
    };
    /**
     * @return {?}
     */
    HtmlCollection.prototype.update = /**
     * @return {?}
     */
    function () {
        for (var i = 0, len = this._collection.length; i < len; i++) {
            this._collection[i].update();
        }
    };
    /**
     * @return {?}
     */
    HtmlCollection.prototype.removeAll = /**
     * @return {?}
     */
    function () {
        while (this._collection.length > 0) {
            /** @type {?} */
            var html = this._collection.pop();
            html.remove();
        }
    };
    /**
     * @param {?} html
     * @return {?}
     */
    HtmlCollection.prototype.contains = /**
     * @param {?} html
     * @return {?}
     */
    function (html) {
        return Cesium.defined(html) && html.collection === this;
    };
    return HtmlCollection;
}());
export { HtmlCollection };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HtmlCollection.prototype._collection;
}
