// angular-cesium/esm2015/lib/cesium-extender/collections/html.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HtmlPrimitive } from '../primitives';
export class HtmlCollection {
    constructor() {
        this._collection = [];
    }
    /**
     * @return {?}
     */
    get length() {
        return this._collection.length;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this._collection[index];
    }
    /**
     * @param {?} options
     * @return {?}
     */
    add(options) {
        /** @type {?} */
        const html = new HtmlPrimitive(options, this);
        this._collection.push(html);
        return html;
    }
    /**
     * @param {?} html
     * @return {?}
     */
    remove(html) {
        /** @type {?} */
        const index = this._collection.indexOf(html);
        if (index === (-1)) {
            return false;
        }
        this._collection[index].remove();
        this._collection.splice(index, 1);
        return true;
    }
    /**
     * @return {?}
     */
    update() {
        for (let i = 0, len = this._collection.length; i < len; i++) {
            this._collection[i].update();
        }
    }
    /**
     * @return {?}
     */
    removeAll() {
        while (this._collection.length > 0) {
            /** @type {?} */
            const html = this._collection.pop();
            html.remove();
        }
    }
    /**
     * @param {?} html
     * @return {?}
     */
    contains(html) {
        return Cesium.defined(html) && html.collection === this;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    HtmlCollection.prototype._collection;
}
