// angular-cesium/esm2015/lib/angular-cesium/services/context-menu/context-menu.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Injectable } from '@angular/core';
import { CesiumEvent } from '../map-events-mananger/consts/cesium-event.enum';
import { PickOptions } from '../map-events-mananger/consts/pickOptions.enum';
/**
 * The Service manages a singleton context menu over the map. It should be initialized with MapEventsManager.
 * The Service allows opening and closing of the context menu and passing data to the context menu inner component.
 *
 * notice, `data` will be injected to your custom menu component into the `data` field in the component.
 * __Usage :__
 * ```
 *  ngOnInit() {
 *   this.clickEvent$ = this.eventsManager.register({ event: CesiumEvent.RIGHT_CLICK, pick: PickOptions.PICK_ONE });
 *   this.clickEvent$.subscribe(result => {
 *    if (result.entities) {
 *      const pickedMarker = result.entities[0];
 *      this.contextMenuService.open(MapContextmenuComponent, pickedMarker.position, {
 *        data: {
 *          myData: data,
 *          onDelete: () => this.delete(pickedMarker.id)
 *        }
 *      });
 *    }
 *   });
 *  }
 *
 *
 *  private delete(id) {
 *    this.mapMenu.close();
 *    this.detailedSiteService.removeMarker(id);
 *  }
 * ```
 */
export class ContextMenuService {
    constructor() {
        this._showContextMenu = false;
        this._contextMenuChangeNotifier = new EventEmitter();
        this._onOpen = new EventEmitter();
        this._onClose = new EventEmitter();
        this._defaultContextMenuOptions = {
            closeOnLeftCLick: true,
            closeOnLeftClickPriority: 10,
        };
    }
    /**
     * @return {?}
     */
    get contextMenuChangeNotifier() {
        return this._contextMenuChangeNotifier;
    }
    /**
     * @return {?}
     */
    get showContextMenu() {
        return this._showContextMenu;
    }
    /**
     * @return {?}
     */
    get options() {
        return this._options;
    }
    /**
     * @return {?}
     */
    get position() {
        return this._position;
    }
    /**
     * @return {?}
     */
    get content() {
        return this._content;
    }
    /**
     * @return {?}
     */
    get onOpen() {
        return this._onOpen;
    }
    /**
     * @return {?}
     */
    get onClose() {
        return this._onClose;
    }
    /**
     * @param {?} mapEventsManager
     * @return {?}
     */
    init(mapEventsManager) {
        this.mapEventsManager = mapEventsManager;
    }
    /**
     * @template D
     * @param {?} contentComponent
     * @param {?} position
     * @param {?=} options
     * @return {?}
     */
    open(contentComponent, position, options = {}) {
        this.close();
        this._content = contentComponent;
        this._position = position;
        this._options = Object.assign({}, this._defaultContextMenuOptions, options);
        this._showContextMenu = true;
        if (this.mapEventsManager && this._options.closeOnLeftCLick) {
            this.leftClickRegistration = this.mapEventsManager.register({
                event: CesiumEvent.LEFT_CLICK,
                pick: PickOptions.NO_PICK,
                priority: this._options.closeOnLeftClickPriority,
            });
            this.leftClickSubscription = this.leftClickRegistration.subscribe(() => {
                this.leftClickSubscription.unsubscribe();
                this.close();
            });
        }
        this._contextMenuChangeNotifier.emit();
        this._onOpen.emit();
    }
    /**
     * @return {?}
     */
    close() {
        this._content = undefined;
        this._position = undefined;
        this._options = undefined;
        this._showContextMenu = false;
        if (this.leftClickRegistration) {
            this.leftClickRegistration.dispose();
            this.leftClickRegistration = undefined;
        }
        if (this.leftClickSubscription) {
            this.leftClickSubscription.unsubscribe();
            this.leftClickSubscription = undefined;
        }
        this._contextMenuChangeNotifier.emit();
        this._onClose.emit();
    }
}
ContextMenuService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._showContextMenu;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._options;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._position;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._content;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype.mapEventsManager;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype.leftClickRegistration;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype.leftClickSubscription;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._contextMenuChangeNotifier;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._onOpen;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._onClose;
    /**
     * @type {?}
     * @private
     */
    ContextMenuService.prototype._defaultContextMenuOptions;
}
